### PR TITLE
KAFKA-3559: Recycle old tasks when possible

### PR DESCRIPTION
**Pull request by enothereska**

### CODE REVIEW

**Comment by enothereska** [comment]
@dguy @guozhangwang please have a look. I confirmed this passes the system tests too.

**Comment by enothereska** [comment]
An unrelated test failed `kafka.integration.PrimitiveApiTest > testPipelinedProduceRequests FAILED`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 1]
```diff
@@ -314,6 +319,42 @@ private void shutdownTasksAndState(final boolean rethrowExceptions) {
...
+    private void suspendTasksAndState(final boolean rethrowExceptions) {
+        log.debug("{} suspendTasksAndState: suspending all active tasks [{}] and standby tasks [{}]", logPrefix,
+            activeTasks.keySet(), standbyTasks.keySet());
+
+        // Commit first as there may be cached records that have not been flushed yet.
+        commitOffsets(rethrowExceptions);
+        // flush state
+        flushAllState(rethrowExceptions);
+        // flush out any extra data sent during close
+        producer.flush();
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:343]: Not sure we need this as the tasks aren't being closed, so there shouldn't be any state to flush - right?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 2]
```diff
@@ -314,6 +319,42 @@ private void shutdownTasksAndState(final boolean rethrowExceptions) {
...
+        try {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I think this `try{..} catch(..)` is identical to the block in `shutdownTasksAndState` - extract to a method to avoid duplication?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 3]
```diff
@@ -92,7 +92,11 @@
     private final Map<TaskId, StandbyTask> standbyTasks;
     private final Map<TopicPartition, StreamTask> activeTasksByPartition;
     private final Map<TopicPartition, StandbyTask> standbyTasksByPartition;
+    private final Map<TopicPartition, StreamTask> suspendedTasksByPartition;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: We have a lot of task related Maps. I wonder if would could encapsulate them behind another Object? Blocks of code like lines 345 - 355 would then be a method on the new Object. Admittedly, i haven't completely thought it through...

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 4]
```diff
@@ -657,7 +698,22 @@ private void addStreamTasks(Collection<TopicPartition> assignment) {
             Set<TopicPartition> partitions = entry.getValue();
 
             try {
-                StreamTask task = createStreamTask(taskId, partitions);
+                StreamTask task = null;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I'd find this easier to follow if it was sth like:

```
StreamTask task = findMatchingSuspendedTask(taskId, partitions);
                if (task != null) {
                    log.debug("{} recycling old task {}", logPrefix, taskId);
                    suspendedTasks.remove(taskId);
                } else {
                    log.debug("{} creating new task {}", logPrefix, taskId);
                    task = createStreamTask(taskId, partitions);
                }
                activeTasks.put(taskId, task);
```

```
private StreamTask findMatchingSuspendedTask(final TaskId taskId, final Set<TopicPartition> partitions) {
        if (suspendedTasks.containsKey(taskId)) {
            final StreamTask task = suspendedTasks.get(taskId);
            if (task.partitions.equals(partitions)) {
                return task;
            }
        }
        return null;
    }
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 5]
```diff
@@ -693,7 +752,22 @@ private void addStandbyTasks() {
         for (Map.Entry<TaskId, Set<TopicPartition>> entry : partitionAssignor.standbyTasks().entrySet()) {
             TaskId taskId = entry.getKey();
             Set<TopicPartition> partitions = entry.getValue();
-            StandbyTask task = createStandbyTask(taskId, partitions);
+            StandbyTask task = null;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: same as comment above about `addStreamTasks`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 6]
```diff
@@ -744,6 +820,34 @@ private void removeStandbyTasks() {
         standbyRecords.clear();
     }
 
+    private void removeSuspendedTasks() {
+        log.info("{} Removing all suspended tasks [{}]", logPrefix, suspendedTasks.keySet());
+        try {
+            // Close task and state manager
+            for (final AbstractTask task : suspendedTasks.values()) {
+                task.close();
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:826]: Do we want to flush the producer after this call? i.e., `task.close` could result in more data being produced.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 7]
```diff
@@ -744,6 +820,34 @@ private void removeStandbyTasks() {
...
+            log.error("{} Failed to remove suspended tasks: ", logPrefix, e);
+        }
+    }
+
+    private void removeSuspendedStandbyTasks() {
+        log.info("{} Removing all suspended standby tasks [{}]", logPrefix, suspendedStandbyTasks.keySet());
+        try {
+            // Close task and state manager
+            for (final AbstractTask task : suspendedStandbyTasks.values()) {
+                task.close();
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:843]: Same as above.

**Review by dguy — CHANGES_REQUESTED** [review]
A few comments/suggestions

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:343]: But if we flush state, we write to the changelogger with the producer, so shouldn't we flush the producer?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 4 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Good suggestion, thanks!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 7 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:843]: Yup thanks!

**Comment by enothereska** [comment]
Not done with changes yet.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 2 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: ok

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: might be an over-kill for now...

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:343]: There is no harm in doing it here, but it also needs to happen after the tasks are closed.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:343]: I agree with @dguy , since during the closure of the topology's processor, user's customized `close()` may send more records.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: `suspendedStandbyTasksByPartition` and `suspendedTasksByPartition` maps are not used anywhere?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 8]
```diff
@@ -303,15 +320,38 @@ private void shutdownTasksAndState(final boolean rethrowExceptions) {
...
+        producer.flush();
+        // remove the changelog partitions from restore consumer
+        unAssignChangeLogPartitions(rethrowExceptions);
+
+        suspendedTasks.clear();
+        suspendedTasks.putAll(activeTasks);
+        suspendedStandbyTasks.putAll(standbyTasks);
+
+        suspendedTasksByPartition.clear();
+        suspendedTasksByPartition.putAll(activeTasksByPartition);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: We still need to update the `prevTasks` which will be used in the rebalance protocol in order to apply heuristics to reassign tasks to those who previously owned them in best-effort.

So if we do not need `suspendedStandbyTasksByPartition` and `suspendedTasksByPartition` (see above comments), probably we can wrap this logic as `updateSuspendedTasks` along with `removeStreamTasks` and `removeStandbyTasks`, and call all three methods in this case.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 9]
```diff
@@ -303,15 +320,38 @@ private void shutdownTasksAndState(final boolean rethrowExceptions) {
...
+     * Similar to shutdownTasksAndState, however does not close the task managers,
+     * in the hope that soon the tasks will be assigned again
+     * @param rethrowExceptions
+     */
+    private void suspendTasksAndState(final boolean rethrowExceptions) {
+        log.debug("{} suspendTasksAndState: suspending all active tasks [{}] and standby tasks [{}]", logPrefix,
+            activeTasks.keySet(), standbyTasks.keySet());
+
+        // Commit first as there may be cached records that have not been flushed yet.
+        commitOffsets(rethrowExceptions);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:337]: We still need to "close" the topology by calling the `close()` call on the processor nodes following the DAG ordering to finish up any user-customizable computing logic, since if the tasks will not be resumed after the rebalance, we need to make sure that they are restarted from a clean state on the other client.

However, some `close()` function implementation maybe not-undoable when the tasks are resumed: think if users decide to close any other external resources in the `close()` call or simply look at `KeyValuePrinterProcessor`. 

So I think we cannot save the cost of closing the topology and re-initialize it: we still need to close the processor nodes, and then upon resuming we should call `init` on the processor nodes again: maybe we can extract the code

```
// initialize the task by initializing all its processor nodes in the topology
        log.info("{} Initializing processor nodes of the topology", logPrefix);
        for (ProcessorNode node : this.topology.processors()) {
            this.currNode = node;
            try {
                node.init(this.processorContext);
            } finally {
                this.currNode = null;
            }
        }
```

in `AbstractTask` with a separate `init` method?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 6 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:826]: Also, I think we should also call `task.flushState` since the flushing has been extracted from the close call. And the ordering of these operations should be
1. `task.close()`
2. `task.flushStore()`
3. `task.closeStateManager()`

Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:826]: Ok

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:343]: Ok

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Good catch.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 8 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Good point thanks.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 10]
```diff
@@ -328,13 +325,44 @@ public void schedule(long interval) {
...
 
+    @Override
+    public void init() {
+        if (prevPartitionQueues.size() > 0) {
+            partitionQueues.clear();
+            partitionQueues.putAll(prevPartitionQueues);
+            this.partitionGroup = new PartitionGroup(partitionQueues, timestampExtractor);
+            prevPartitionQueues.clear();
+        }
+        if (prevConsumedOffsets.size() > 0) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Why do we want to re-create the `partitionGroup` with the previous partition queues? Could we just keep the old `partitionGroup` of the task as is? And also I think we do not need to keep `prevConsumedOffsets` as the `consumedOffsets` can always be initialized by the prev committed offsets which we will still execute for pausing the task.

Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: @guozhangwang the problem is that task.close clears the partitionGroup, so I'm trying to keep it around this way.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Hmm, could we extract the closure of the nodes only into a separate function, like `closeTopology` and rename newly added `init` to `initTopology`?

Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Ok will do thanks.

**Comment by guozhangwang** [comment]
Also could you paste the time spent on `KStreamRepartitionJoinTest` with your local run before and after the patch, so that we can have some information on how much savings it has?

**Comment by enothereska** [comment]
@guozhangwang before the patch, average of 3 runs is 47 seconds. After the patch it is 44 seconds. I'm not sure the optimization is worth it.

**Comment by guozhangwang** [comment]
@enothereska could we profile which part of the `onPartitionAssigned` and `onPartitionRevoked` functions are the bottleneck then?

**Comment by enothereska** [comment]
@guozhangwang turns out I had a bit of logging going on. With no logging the numbers from 5 runs of shouldCorrectlyRepartitionOnJoinOperations are:
- with shutdown: 22, 22, 22, 21, 21 seconds (avg 21.6)
- with suspend:   17, 17, 15, 14, 14 seconds (avg 15.4)

So that's around 28% improvement.

**Comment by guozhangwang** [comment]
Thanks @enothereska the improvement looks significant for the worthwhileness of the optimization. Could we close / reopen this PR to let Jenkins trigger another build since the reported errors are new to me but I cannot reproduce it locally.

**Comment by enothereska** [comment]
@guozhangwang there is definitely a bug there and we shouldn't merge until it's fixed. Still chasing, it's tricky to reproduce.

**Comment by enothereska** [comment]
System tests (smoke and bounce tests) passed.

**Comment by enothereska** [comment]
Unrelated jenkins environment failure

**Comment by enothereska** [comment]
Same jenkins failure. Looks like something is wrong with Jenkins.

**Comment by guozhangwang** [comment]
@enothereska Could you try retrigger jenkins again? https://github.com/apache/kafka/pull/2012 's jenkins succeeded.

**Comment by enothereska** [comment]
@guozhangwang worked, thanks.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java [hunk 11]
```diff
@@ -170,4 +170,10 @@ public void close() {
         queuesByTime.clear();
         partitionQueues.clear();
     }
+
+    public void clear() {
+        for (RecordQueue queue : partitionQueues.values()) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:176]: We also need to clear the `queuesByTime`.

**Comment by guozhangwang** [comment]
Could you rebase as well?

**Comment by enothereska** [comment]
KStreamAggregationIntegrationTest > shouldGroupByKey[1] FAILED I haven't seen this failure before, will need investigating.

**Comment by enothereska** [comment]
I opened a JIRA for this one (https://issues.apache.org/jira/browse/KAFKA-4359), it's not related to this PR and I'll need to clean up multiple tests. Will also commit a fix just for shouldGroupByKey for this PR.

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 12]
```diff
@@ -43,10 +43,6 @@
 import org.junit.Before;
 import org.junit.ClassRule;
 import org.junit.Test;
-import org.junit.runner.RunWith;
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:46]: Is this change intentional? Why we want to remove the parameters?

Comment (enothereska) [reply streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:46]: Because KStreamAggregationIntegrationTest was never supposed to run with caching, we have the other test KStreamAggregationDedupIntegrationTest that was supposed to contain the same tests but with caching.

**Comment by enothereska** [comment]
Failure in Kafka unrelated to PR, but stopped rest of tests.

**Comment by enothereska** [comment]
Usual transient reset tool failure, all else passed.

**Comment by guozhangwang** [comment]
Merged to trunk.
